obspy/clients/syngine/client.py

```diff
@@ -0,0 +1,602 @@
+                        (time_arguments, obspy.UTCDateTime)):
+            for key in keys:
+                try:
+                    value = kwargs[key]
+                except KeyError:
+                    continue
+                if value is None:
+                    continue
+                value = t(value)
+                # String arguments are stripped and empty strings are not
+                # allowed.
+                if t is str:
+                    value = value.strip()
+                    if not value:
+                        raise ValueError("String argument '%s' must not be "
+                                         "an empty string." % key)
+                params[key] = t(value)
+
+        # These can be absolute times, relative times or phase relative times.
+        # jo
+        temporal_bounds = ["starttime", "endtime"]
+        for key in temporal_bounds:
+            try:
+                value = kwargs[key]
+            except KeyError:
+                continue
+            if value is None:
+                continue
+            # If a number, convert to a float.
+            elif isinstance(value, (int, float)):
+                value = float(value)
+            # If a string like object, attempt to parse it to a datetime
+            # object, otherwise assume its a phase-relative time and let the
+            # Syngine service deal with the erorr handling.
+            elif isinstance(value, (str, native_str)):
+                try:
+                    value = obspy.UTCDateTime(value)
+                except:
+                    value = (value)
+            # Last but not least just try to pass it to the datetime
+            # constructor without catching the error.
+            else:
+                value = obspy.UTCDateTime(value)
+            params[key] = native_str(value)
+
+        # These all have to be lists of floats. Otherwise it fails.
+        source_mecs = ["sourcemomenttensor",
+                       "sourcedoublecouple",
+                       "sourceforce"]
+        for key in source_mecs:
+            try:
+                value = kwargs[key]
+            except KeyError:
+                continue
+            if value is None:
+                continue
+            value = ",".join(["%g" % float(_i) for _i in value])
+            params[key] = value
+
+        return params
+
+    def __read_to_stream(self, r):
+        with io.BytesIO(r.content) as buf:
+            # First try to read the file in a normal way, otherwise assume
+            # its a saczip file.
+            try:
+                st = obspy.read(buf)
+            except:
+                st = obspy.Stream()
+                zip_obj = zipfile.ZipFile(io.BytesIO(r.content))
+                for name in zip_obj.namelist():
+                    # Skip the log file.
+                    if name.lower() == "syngine.log":
+                        continue
+                    st += obspy.read(io.BytesIO(zip_obj.read(name)))
```

Comment:
I don't understand why the top-level `BytesIO` object gets put in a `with` context but this one doesn't.